safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,182 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csp
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Directive is the name of a single CSP directive.
+type Directive string
+
+const (
+	DirectiveScriptSrc Directive = "script-src"
+	DirectiveStyleSrc  Directive = "style-src"
+	DirectiveObjectSrc Directive = "object-src"
+	DirectiveBaseURI   Directive = "base-uri"
+	DirectiveReportURI Directive = "report-uri"
+)
+
+const (
+	ValueHTTPS         = "https:"
+	ValueHTTP          = "http:"
+	ValueUnsafeEval    = "'unsafe-eval'"
+	ValueUnsafeInline  = "'unsafe-inline'"
+	ValueNone          = "'none'"
+	ValueStrictDynamic = "'strict-dynamic'"
+)
+
+// PolicyDirective contains a single CSP directive.
+type PolicyDirective struct {
+	dir       Directive
+	vals      []string
+	nonces    bool
+	lastNonce string
+	// readRand is used for dependency injection in tests.
+	readRand func([]byte) (int, error)
+}
+
+// NewPolicyDirective creates a new PolicyDirective given the directive, the
+// values and whether or not a nonce directive should be added.
+func NewPolicyDirective(d Directive, values []string, nonces bool) *PolicyDirective {
+	return &PolicyDirective{
+		dir:      d,
+		vals:     values,
+		nonces:   nonces,
+		readRand: rand.Read,
+	}
+}
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func (pd *PolicyDirective) generateNonce() {
+	b := make([]byte, nonceSize)
+	_, err := pd.readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	pd.lastNonce = base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy, containing many directives.
+type Policy struct {
+	Directives []*PolicyDirective
+}
+
+// NewPolicy creates a new strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// TODO: maybe reportURI should be safehttp.URL?
+func NewPolicy(reportURI string) *Policy {
+	return &Policy{
+		Directives: []*PolicyDirective{
+			NewPolicyDirective(DirectiveObjectSrc, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveScriptSrc, []string{
+				ValueUnsafeInline,
+				ValueUnsafeEval,
+				ValueStrictDynamic,
+				ValueHTTPS,
+				ValueHTTP,
+			}, true),
+			NewPolicyDirective(DirectiveBaseURI, []string{ValueNone}, false),
+			NewPolicyDirective(DirectiveReportURI, []string{reportURI}, false),
+		},
+	}
+}
+
+// Serialize serializes this policy for use in a Content-Security-Policy header
+// or in a Content-Security-Policy-Report-Only header. The nonces generated for
+// each directive is also returned.
+func (p Policy) Serialize() (csp string, nonces map[Directive]string) {
+	nonces = make(map[Directive]string)
+	b := strings.Builder{}
+
+	for i, d := range p.Directives {
+		if i != 0 {
```

Comment:
If this is your intention, maybe just create a `[]string` and use `strings.Join`?